#167165057 User Can Sign Out

tolulope-od opened the pull request:
### What does this PR do?

Creates functionality for users to sign out
Creates an authentication middleware
Creates a server response module
Add pretest script
Changes blacklist table token column data type
Adds database service to Travis CI

#### Description of Task to be completed?
When a user signs out of the app (drops token), then the user should not be able to access any route protected by the token.

User should be able to sign out via the route: `api/v1/auth/logout`
 
#### How should this be manually tested?
After cloning this repo, `cd` into it and do the following:

```bash
# fetch this branch
$ git fetch origin ft-user-sign-out-167165057

# install dependencies
$ npm install

# run unit tests
$ npm test

# start the development server
$ npm run server
```

When the server is running, using Postman, make a `POST` request to the route stated above

#### Any background context you want to provide?
Given the fact that when a JSON Web Token is generated, it cannot be unsigned, the way this feature was implemented was to have a table that stores blacklisted tokens. This table is populated when a user hits the logout route, hence this route is protected by an Authentication middleware that checks for a token.

#### What are the relevant pivotal tracker stories?
[#167165057](https://www.pivotaltracker.com/story/show/167165057)

#### Screenshots (if appropriate)
*Successful logout request*

<img width="1285" alt="Screenshot 2019-07-25 at 16 08 01" src="https://user-images.githubusercontent.com/42325628/61899454-64afe800-af13-11e9-8aa3-303b0d1f495c.png">

tolulope-od committed:
feature(auth): user can sign out

Create functionality for user to sign out
Create authentication middleware
Create server response module
Add pretest script
Change blacklist table model column
Add database service to travis CI

[Starts #167165057]

tolulope-od committed:
change travis config

OdunayoOkebunmi reviewed (commented):
Kindly include screenshots.

OdunayoOkebunmi commented on controllers/AuthController.js:
```diff
@@ -0,0 +1,89 @@
+   * @static
+   * @param {object} req express request object
+   * @param {object} res express response object
+   * @returns {object} returns user data
```
kindly separate your JSDoc block tags

tolulope-od replied:
I pulled in files from someone else's branch and must have missed this but I will make these changes. Thank you

tolulope-od replied:
Done

tolulope-od commented:
> Kindly include screenshots.

Done.

mekzy-o reviewed (approved):
Good Job Tolu. Though your file changes were much. Keep it less next time.

mazma1 commented on .env:
```diff
@@ -0,0 +1,3 @@
+DATABASE_URL=postgres://tolulopeodueke@localhost:5432/mazus_ah_backend
+TEST_DATABASE_URL=postgres://tolulopeodueke@localhost:5432/mazus_ah_backend_test
+SECRET_KEY=master
```
What is `.env` doing on Github again???

tolulope-od replied:
I didn't know this was added again, I'll remove it. Thanks

tolulope-od replied:
Done

mazma1 commented on database/models/user.js:
```diff
@@ -3,92 +3,95 @@
       as: 'profile',
       onDelete: 'CASCADE',
-      onUpdate: 'CASCADE',
+      onUpdate: 'CASCADE'
```
Please put back the trailing commas you removed from this file as they are important.

Note: Don't go making changes to files that have nothing to do with what you are working on.

tolulope-od replied:
Done

Tunji17 reviewed (commented):
Good job @tolulope-od but is there a reason why the normal migrate script is crossing env to test before running migrate?

tolulope-od commented:
> Good job @tolulope-od but is there a reason why the normal migrate script is crossing env to test before running migrate?

Thanks for pointing that out. That was an oversight. I'll fix that now

tolulope-od committed:
chore(refactor): implement feedback from PR

Add better test assertions
Fix trailing comma rule in ESLint
Fix files missing trailing commas

[Finishes #167165057]

mazma1 commented on database/migrations/20190723105203-create-user.js:
```diff
@@ -45,7 +45,7 @@ module.exports = {
+    },
   }),
-  down: (queryInterface, Sequelize) => queryInterface.dropTable('Users')
+  down: (queryInterface, Sequelize) => queryInterface.dropTable('Users'),
```
Please enforce trailing commas for only multiline array and object statements. It doesn't apply here

tolulope-od replied:
ESLint flags it when I do that.
